### DISCUSSION TITLE
3.0 中 sandbox 里获取 document.body 返回的是容器节点 而不是真正的 document.body

### DISCUSSION THREAD

**Discussion by ZenDay** [umijs/qiankun#2905] [Q&A]
![image](https://github.com/umijs/qiankun/assets/7631962/87d09345-9b47-4e86-ae51-d051344fd934)

留意到 sandbox 中读取 document.body 时，实际上通过代理来返回了容器节点，而不是真正document.body。

但这种做法在很多库中会存在问题。

如：
1. 如果监听全局的键盘事件时使用了`document.body.addEventListener('keydown', listener)`，则会导致只要我的焦点没有点中对应的元素，就没办法监听键盘事件；
2. 在 antdv 中，popover 和 tooltips 的生成方式是在 document.body 后插入div元素来做元素的相对定位。但如果返回的是容器节点，会导致我的全局 popover/tooltips 没办法在全局范围内做定位和zIndex，而是依赖容器节点的定位和zIndex。

想咨询一下有什么特殊原因一定需要document.body要返回容器节点，又有什么方法可以在实际的业务代码中禁用这一层代理呢？

**Comment by bravepg** [comment]
如果自己使用的时候，有个绕开的方式，用 `document.documentElement.addEventListener` 来替代 `document.body.addEventListener`。

**Comment by kuitos** [comment]
3.0 提供了自己复写沙箱里全局变量的方案，比如将沙箱里的 document 设置成宿主环境的 document，而不是现在的代理对象。

**Reply by bravepg** [reply]
如何复写？
